rcl/src/rcl/logging_rosout.c

```diff
@@ -0,0 +1,236 @@
+  if (NULL == logger_name) {
+    RCL_SET_ERROR_MSG("Logger name was null.");
+    return RCL_RET_ERROR;
+  }
+  if (rcutils_hash_map_key_exists(&__logger_map, &logger_name)) {
+    RCL_SET_ERROR_MSG("Logger already initialized for node.");
+    return RCL_RET_ALREADY_INIT;
+  }
+
+  // Create a new Log message publisher on the node
+  const rosidl_message_type_support_t * type_support =
+    rosidl_typesupport_c__get_message_type_support_handle__rcl_interfaces__msg__Log();
+  rcl_publisher_options_t options = rcl_publisher_get_default_options();
+  new_entry.publisher = rcl_get_zero_initialized_publisher();
+  status =
+    rcl_publisher_init(&new_entry.publisher, node, type_support, ROSOUT_TOPIC_NAME, &options);
+
+  // Add the new publisher to the map
+  if (RCL_RET_OK == status) {
+    new_entry.node = node;
+    status = rcutils_hash_map_set(&__logger_map, &logger_name, &new_entry);
+    if (RCL_RET_OK != status) {
+      RCL_SET_ERROR_MSG("Failed to add publisher to map.");
+      // We failed to add to the map so destroy the publisher that we created
+      rcl_ret_t fini_status = rcl_publisher_fini(&new_entry.publisher, new_entry.node);
+      // ignore the return status in favor of the failure from set
+      RCL_UNUSED(fini_status);
+    }
+  }
+
+  return status;
+}
+
+rcl_ret_t rcl_logging_rosout_fini_publisher_for_node(
+  rcl_node_t * node)
+{
+  RCL_LOGGING_ROSOUT_VERIFY_INITIALIZED
+  rosout_map_entry_t entry;
+  const char * logger_name = NULL;
+  rcl_ret_t status = RCL_RET_OK;
+
+  // Verify input and make sure it's initialized
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+  logger_name = rcl_node_get_logger_name(node);
+  if (NULL == logger_name) {
+    return RCL_RET_ERROR;
+  }
+  if (!rcutils_hash_map_key_exists(&__logger_map, &logger_name)) {
+    return RCL_RET_NOT_INIT;
+  }
+
+  // fini the publisher and remove the entry from the map
+  status = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
+  if (RCL_RET_OK == status) {
+    status = rcl_publisher_fini(&entry.publisher, entry.node);
+  }
+  if (RCL_RET_OK == status) {
+    status = rcutils_hash_map_unset(&__logger_map, &logger_name);
+  }
+
+  return status;
+}
+
+void rcl_logging_rosout_output_handler(
+  const rcutils_log_location_t * location,
+  int severity, const char * name, rcutils_time_point_value_t timestamp,
+  const char * format, va_list * args)
+{
+  rosout_map_entry_t entry;
+  rcl_ret_t status = RCL_RET_OK;
+  if (!__is_initialized) {
+    return;
+  }
+  status = rcutils_hash_map_get(&__logger_map, &name, &entry);
+  if (RCL_RET_OK == status) {
+    char msg_buf[1024] = "";
+    rcutils_char_array_t msg_array = {
+      .buffer = msg_buf,
+      .owns_buffer = false,
+      .buffer_length = 0u,
+      .buffer_capacity = sizeof(msg_buf),
+      .allocator = __rosout_allocator
+    };
+
+    va_list args_clone;
+    va_copy(args_clone, *args);
+    status = rcutils_char_array_vsprintf(&msg_array, format, args_clone);
```

Comment:
Maybe we should check the return value. If there is an error we can print an error and return early.

Reply:
Checked for error and printing, but didn't return early so that fini could be called on resources just in case.